doc/proposals/scorecard-custom-tests.md

```diff
@@ -0,0 +1,403 @@
+
+Test assertions are expected to be provided by the scorecard end user in YAML format.  The path to those assertions would be configured within the scorecard configuration file or command line flags.  Test assertions are related and physically organized on disk to match Test Steps.
+
+#### Test Images
+
+Complex tests would require a test container image, that image would be provided by the end user.  The test image would have some common interfaces such as where the scorecard configuration files (e.g. CRs) would be mounted within the container, and how the pass/fail is represented (e.g. Failed or Successful).  There would also be an expected format of the test image log output so that scorecard could process the output log if necessary.
+
+The test image would  adhere to the following common interface:
+
+* Output would be in v1alpha2 ScorecardTestResult format, this is required for the scorecard to parse the test image log output and report the test results to the scorecard user, all scorecard tests would require this output format.
+
+ * Input would be a ConfigMap holding the scorecard configuration such as the CR being tested.  The ConfigMap would follow a naming convention of scorecard-custom-test and would include the CR being tested.  The custom test image could either mount the ConfigMap within its Dockerfile or read it dynamically using an API (e.g.  client-go).
+
+Test image qualified names are specified in Test Steps.
+
+### Custom Test Format
+Let's describe what a test would look like in the proposed design.
+
+#### State Comparison Test Format
+
+Custom tests can be written in the form of assertions that are built using YAML.  Assertions are simple tests to see if a particular Kubernetes resource is in a particular state after some period of time.  Examples of an assertion might be a number of requested pods that result from the creation of a particular Custom Resource (CR).  The scorecard executes the assertion, watching for a specific state, and returns a ScorecardTestResult upon timeout.
+
+An example of a State Comparison test (assertion) might look as follows:
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-deployment
+status:
+  readyReplicas: 3
+```
+
+In the above example, this assertion is evaluated as being true or not, this assumes that a Deployment, named example-deployment, with 3 replicas is present on the Kube cluster for example.   The scorecard would execute this assertion as a part of a custom test, the assertion is a YAML manifest.
+
+Test setup steps for a test of this type can be defined using YAML such as:
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-deployment
+  labels:
+    app: nginx
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: nginx
+...
+```
```

Comment:
Btw I loved 👍 the idea of this Assert tests. They show very easy to be used which would be a great improvement in user experience.

Reply:
yep, I think end users would like to write tests in the manner, that being to construct assert/test step YAML, store them on-disk and run a simple command line to execute those tests.